Fix ImageMap default values

Prokyonn opened the pull request:
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #issuenum
| Related issues/PRs | #issuenum
| License | MIT
| Documentation PR | sulu/sulu-docs#prnum

#### What's in this PR?

Fixes default values for ImageMap similar to https://github.com/sulu/sulu/pull/5576

Furthermore some changes in `FieldBlocks` have been made to be more consistent with other components.

#### Why?

Default values should work as expected.

danrot commented on src/Sulu/Bundle/MediaBundle/Resources/js/containers/Form/fields/ImageMap/ImageMap.js:
```diff
@@ -10,10 +12,33 @@ import FieldRenderer from './FieldRenderer';
+        this.setValue(this.props.value);
+    }
+
+    componentDidUpdate(prevProps: FieldTypeProps<Value>): * {
```
Is the definition of the return type value necessary here? :thinking:

danrot reviewed (changes requested):
This also needs a rebase, right? You probably have to update the code I've commited in #5593 too.

Prokyonn commented:
> This also needs a rebase, right? You probably have to update the code I've commited in #5593 too.

rebased and removed the unnecessary return type

danrot commented:
The block settings are not working anymore with this PR (if I confirm the overlay the changes are immediately lost).

Prokyonn commented on src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js:
```diff
@@ -372,17 +368,20 @@ class FieldBlocks extends React.Component<FieldTypeProps<Array<BlockEntry>>> {
             return;
         }
 
+        value[blockSettingsOpen] = {...value[blockSettingsOpen], [SETTINGS_KEY]: blockSettingsFormStore.data};
```
Added the update via the `setValue` method. But as I understand it, we can change the `onChange` call because now the settings are already updated in the `newValues` and thus we only need to make a new object to avoid passing the same reference with `{...newValues[blockSettingsOpen]}` or am I missing something?

danrot replied:
Not sure what you are talking about, but I already added another review comment, maybe that answers your question :see_no_evil:

Prokyonn replied:
Yeah I misunderstood it 🙈 . Your snipped helped a lot.

danrot commented on src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js:
```diff
@@ -372,18 +368,24 @@ class FieldBlocks extends React.Component<FieldTypeProps<Array<BlockEntry>>> {
+            ...newValues.slice(0, blockSettingsOpen),
+            {...newValues[blockSettingsOpen]},
+            ...newValues.slice(blockSettingsOpen + 1),
         ]);
```
Now the `toJS` and the `onChange` call create a new array, which kind of is doubling the work :see_no_evil: I think I would have either just called `onChange(newValues)`, or would have rewritten that differently (that would be my favourite approach):

```javascript
const newValues = [
    ...oldValues.slice(0, blockSettingsOpen),
    {...oldValues[blockSettingsOpen], [SETTINGS_KEY]: blockSettingsFormStore.data},
    ...oldValues.slice(blockSettingsOpen +1),
];

this.setValue(newValues);
onChange(newValues);
```

This would make the call to `toJS` and `jsonpointer.set` obsolete.
